File: 1190. Reverse Substrings Between Each Pair of Parentheses.cpp
class Solution {
public:
        
    string reverseParentheses(string s) {
        stack<int> stk;
        for (int i=0; i < s.size(); ++i) 
            if (s[i]=='(')
                stk.push(i);
            else if (s[i]==')') {
                int p = stk.top(); stk.pop();
                reverse(s.begin()+p+1, s.begin()+i);
            }
        string ans;
        for (auto c : s) 
            if (isalpha(c)) ans.push_back(c);
        return ans;
    }
};

File: 1190. Reverse Substrings Between Each Pair of Parentheses.py
class Solution:
    def reverseParentheses(self, s: str) -> str:
        n = len(s)
        pr, stk = [0] * n, []
        for i in range(n):
            if s[i] == '(':
                stk.append(i)
            elif s[i] == ')':
                j = stk.pop()
                pr[i], pr[j] = j, i
        i, d, ans = 0, 1, ""
        while i < n:
            if s[i].isalpha():
                ans += s[i]
            else:
                i, d = pr[i], -d
            i += d
        return ans
                
                
        